### PR TITLE
Zipkin

### PR DESCRIPTION
**Pull request by antmat**
Store trace in upstream in case upstream is a client upstream in order to group client calls with same span_id

### CODE REVIEW

**Comment by antmat** [comment]
WIP, do not merge at the moment.

File: include/cocaine/trace/trace.hpp [hunk 1]
```diff
@@ -58,12 +61,21 @@ class trace_t {
...
         service_name(_service_name)
     {
+        if(parent_id == reserved_value ||
+           span_id == reserved_value ||
+           trace_id == reserved_value ||
+           // Partially empty trace - trace_id without span_id, vise versa or parent_id without trace_id
+           (span_id == zero_value != trace_id == zero_value) ||
+           (trace_id == zero_value && parent_id != zero_value)
+        ) {
+            throw cocaine::error_t("Invalid trace parameters: %lli %lli %lli", trace_id, span_id, parent_id);
```
Comment (3Hren) [inline include/cocaine/trace/trace.hpp]: %llu

File: include/cocaine/trace/trace.hpp [hunk 2]
```diff
@@ -58,12 +61,21 @@ class trace_t {
...
         span_id(_span_id),
         parent_id(_parent_id),
-        parent_parent_id(),
+        parent_parent_id(reserved_value),
         start_time_us(),
         last_time_us(),
         rpc_name(_rpc_name),
         service_name(_service_name)
     {
+        if(parent_id == reserved_value ||
```
Comment (3Hren) [inline include/cocaine/trace/trace.hpp]: I'd rather create a separate class for tracing numbers, where checks occurs.

File: include/cocaine/trace/trace.hpp [hunk 3]
```diff
@@ -101,22 +115,28 @@ class trace_t {
...
+        return trace_id == zero_value;
     }
 
     void
     pop() {
-        assert(parent_id != 0);
+        if(empty()) {
+            return;
+        }
+        assert(parent_id != zero_value);
```
Comment (3Hren) [inline include/cocaine/trace/trace.hpp]: Let's use BOOST_ASSERT, it makes an expression MORE VISIBLE.

File: include/cocaine/trace/trace.hpp [hunk 4]
```diff
@@ -101,22 +115,28 @@ class trace_t {
...
         span_id = parent_id;
         parent_id = parent_parent_id;
         rpc_name = parent_rpc_name;
         parent_rpc_name.reset();
-        parent_parent_id = 0;
+        parent_parent_id = reserved_value;
```
Comment (3Hren) [inline include/cocaine/trace/trace.hpp]: grandparent?

File: include/cocaine/trace/trace.hpp [hunk 5]
```diff
@@ -38,11 +40,14 @@ class trace_t {
     class restore_scope_t;
     class push_scope_t;
 
+    static constexpr uint64_t reserved_value = -1;
```
Comment (3Hren) [inline include/cocaine/trace/trace.hpp]: Reserved for what?

File: src/session.cpp [hunk 6]
```diff
@@ -321,13 +345,14 @@ session_t::pull() {
 
 void
 session_t::push(encoder_t::message_type&& message) {
+    COCAINE_LOG_TRACE(log, "Pushing message to session");
```
Comment (3Hren) [inline src/session.cpp]: p

**Comment by 3Hren** [comment]
LAGTM

**Pull request merged**